integration_tests/src/main/python/sort_test.py

```diff
@@ -105,9 +105,11 @@ def test_single_nested_orderby_with_limit(data_gen, order):
 @pytest.mark.parametrize('data_gen', orderable_gens + orderable_not_null_gen, ids=idfn)
 @pytest.mark.parametrize('order', [f.col('a').asc(), f.col('a').asc_nulls_last(), f.col('a').desc(), f.col('a').desc_nulls_first()], ids=idfn)
 def test_single_sort_in_part(data_gen, order):
-    assert_gpu_and_cpu_are_equal_collect(
-            lambda spark : unary_op_df(spark, data_gen).sortWithinPartitions(order),
-            conf = allow_negative_scale_of_decimal_conf)
+    assert_gpu_and_cpu_are_equal_collect_debug(
```

Comment:
It would be nice to make this easier to use, as I suspect we'll have other instances in the future where we'll want to debug-dump a dataframe like this.  The issue with the way it's implemented now is that I can't just change `assert_gpu_and_cpu_are_equal_collect` to `assert_gpu_and_cpu_are_equal_collect_debug` and walk away. I need to also provide a result function and hope I coded it the same way it used to work with `assert_gpu_and_cpu_are_equal_collect` did.

If we added a new `COLLECT_WITH_DEBUG` collect type and corresponding code to dump after the test's dataframe call is made then I think we could make it so tacking on `_debug` to the assert call is all that's required to debug it.

For this specific case where you're wanting to dump the dataframe after the unary_op but before the sort, we actually don't need a new assert method at all.  We could write it as something like this:
```python
def test_single_sort_in_part(data_gen, order):
    def doit(spark):
        df = unary_op_df(spark, data_gen)
        data_gen.debug_df(df)
        return df.sortWithinPartitions(order)
    assert_gpu_and_cpu_are_equal_collect(doit, conf = allow_negative_scale_of_decimal_conf)
```

Reply:
> If we added a new COLLECT_WITH_DEBUG collect type and corresponding code to dump after the test's dataframe call is made then I think we could make it so tacking on _debug to the assert call is all that's required to debug it.

The issue is that what we want to capture is the source dataframe, not the result dataframe. When assert* is called, we already created the result dataframe. There already exists a `COLLECT_WITH_DATAFRAME`, but if I understand correctly, the dataframe that is captured is the one after `sortWithinPartitions` in our case. I agree it can be made easier to use, thought of some options:

1. Just have the `data_gen` instance regenerate. It should create a new Random variable from the same seed, the result should be exactly the same. Need a way to capture which `data_gen` (which is not consistent, i.e. there's a `three_col_df` for example, that can take different `data_gen` intances)

2. Could record generated dataframes in temporary views or some other variable (thought of within `data_gen`). This seems like a bad idea creating side effects.

3. It does seem that a general recipe is: we make a dataframe, then we perform an action on it. The API for this _debug assert was somewhat used in the _sql asserts, where we create a temp view and give it a name, so it can be acted on. I wonder if we really ought to write test in this way?